source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -0,0 +1,264 @@
+  if (!SPIFFEValidator::certificatePrecheck(&leaf_cert)) {
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(Envoy::Ssl::ClientValidationStatus::Failed);
+    }
+    stats_.fail_verify_error_.inc();
+    return 0;
+  }
+
+  auto trust_bundle = getTrustBundleStore(&leaf_cert);
+  if (!trust_bundle) {
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(Envoy::Ssl::ClientValidationStatus::Failed);
+    }
+    stats_.fail_verify_error_.inc();
+    return 0;
+  }
+
+  // set the trust bundle's certificate store on the context, and do the verification
+  store_ctx->ctx = trust_bundle;
+  auto ret = X509_verify_cert(store_ctx);
+  if (ssl_extended_info) {
+    ssl_extended_info->setCertificateValidationStatus(
+        ret == 1 ? Envoy::Ssl::ClientValidationStatus::Validated
+                 : Envoy::Ssl::ClientValidationStatus::Failed);
+  }
+  if (!ret) {
+    stats_.fail_verify_error_.inc();
+  }
+  return ret;
+}
+
+X509_STORE* SPIFFEValidator::getTrustBundleStore(X509* leaf_cert) {
+  bssl::UniquePtr<GENERAL_NAMES> san_names(static_cast<GENERAL_NAMES*>(
+      X509_get_ext_d2i(leaf_cert, NID_subject_alt_name, nullptr, nullptr)));
+  if (!san_names) {
+    return nullptr;
+  }
+
+  std::string trust_domain;
+  for (const GENERAL_NAME* general_name : san_names.get()) {
+    const std::string san = Utility::generalNameAsString(general_name);
+    trust_domain = SPIFFEValidator::extractTrustDomain(san);
+    // we can assume that valid SVID has only one san
+    break;
+  }
+
+  if (trust_domain.empty()) {
+    return nullptr;
+  }
+
+  auto target_store = trust_bundle_stores_.find(trust_domain);
+  return target_store != trust_bundle_stores_.end() ? target_store->second.get() : nullptr;
+}
+
+bool SPIFFEValidator::certificatePrecheck(X509* leaf_cert) {
+  // Check basic constrains and key usage
+  // https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md#52-leaf-validation
+  auto ext = X509_get_extension_flags(leaf_cert);
+  if (ext & EXFLAG_CA) {
+    return false;
+  }
+
+  auto us = X509_get_key_usage(leaf_cert);
+  return (us & (KU_CRL_SIGN | KU_KEY_CERT_SIGN)) == 0;
+}
+
+std::string SPIFFEValidator::extractTrustDomain(const std::string& san) {
+  static const std::string prefix = "spiffe://";
+  size_t begin = 0;
+  for (; begin < san.size() && begin < prefix.size(); begin++) {
+    if (prefix[begin] != san[begin]) {
+      break;
+    }
+  }
+
+  if (begin != prefix.size()) {
+    return "";
+  }
+
+  size_t end = begin;
+  for (; end < san.size(); end++) {
+    if (san[end] == '/') {
+      return san.substr(begin, end - begin);
+    }
+  }
+  return "";
+}
+
+size_t SPIFFEValidator::daysUntilFirstCertExpires() const {
+  if (ca_certs_.empty()) {
+    return 0;
+  }
+  // TODO(mathetake): with the current interface, we cannot pass the multiple cert information.
```

Comment:
I think the interface is adequate for this. The correct behavior is to loop over the certs, and return the minimum value of all of them.